app/src/main/java/com/example/drawingactivity/GradientView.kt

```diff
@@ -0,0 +1,160 @@
+package com.example.drawingactivity
+
+import android.annotation.SuppressLint
+import android.content.Context
+import android.graphics.*
+import android.graphics.Color.BLACK
+import android.graphics.Color.HSVToColor
+import android.util.AttributeSet
+import android.view.MotionEvent
+import android.view.View
+import kotlinx.coroutines.*
+
+@SuppressLint("ClickableViewAccessibility")
+class GradientView @JvmOverloads constructor(
+    context: Context,
+    attrs: AttributeSet? = null,
+    defStyle: Int = 0
+) : View(context, attrs, defStyle) {
+
+    var onSatValChanged: ((Float, Float) -> Unit)? = null
+    private var hue = 359
+    private var shape: Rect? = null
+    private var dx = (width * 1f - 15f)
+    private var dy = 5f
+    private val selectPaint = Paint().apply {
+        strokeWidth = 3f
+        color = BLACK
+        style = Paint.Style.STROKE
+    }
+
+    init {
+        setOnTouchListener { _, mEvent ->
+            when (mEvent.action) {
+                MotionEvent.ACTION_DOWN -> {
+                    dx = mEvent.x.coerceIn(0f, width * 1f)
+                    dy = mEvent.y.coerceIn(0f, height * 1f)
+                    if (dy > height / 2) {
+                        selectPaint.color = Color.WHITE
+                    } else selectPaint.color = BLACK
+                    onSatValChanged?.invoke(dx / width, dy / height)
+                    invalidate()
+                }
+                MotionEvent.ACTION_MOVE, MotionEvent.ACTION_CANCEL -> {
+                    dx = mEvent.x.coerceIn(0f, width * 1f)
+                    dy = mEvent.y.coerceIn(0f, height * 1f)
+                    if (dy > height / 2) {
+                        selectPaint.color = Color.WHITE
+                    } else selectPaint.color = BLACK
+                    onSatValChanged?.invoke(dx / width, dy / height)
+                    invalidate()
+                }
+            }
+            true
+        }
+    }
+
+    @OptIn(DelicateCoroutinesApi::class)
+    private fun generateAllBitmaps() {
+        if (!arrayIsFull) {
+            GlobalScope.launch {
```

Comment:
I imagine theres a better scope to use for this, My guess is Nic will tell us to pass this to the view model or something similar to have it run on the viewModel scope. Not suggesting a change here just giving the thought.

Reply:
I remember him mentioning that Default was _supposed_ to be faster for whatever reason for the dispatcher, but yeah for the launching I am curious if we would want to use a different scope as well.